File: ios/cs/AbrevvaCodingStation.swift
import Foundation
import AbrevvaSDK

@objc(AbrevvaCodingStation)
public class AbrevvaCodingStation: NSObject {
    private let codingStation = CodingStation()
    private var mqttConnectionOptions: MqttConnectionOptions?

    @objc func register(
        _ options: NSDictionary, resolver resolve: @escaping RCTPromiseResolveBlock,
        rejecter reject: @escaping RCTPromiseRejectBlock
    ) {
        guard let args = options as? [String: Any] else {
            return reject("Failed to convert NSDictionary to Swift dictionary", nil, nil)
        }

        guard let url = URL(string: args["url"] as? String ?? "") else {
            return reject( "register(): failed to create URL", nil, nil)
        }
        let clientId = args["clientId"] as? String ?? ""
        let username = args["username"] as? String ?? ""
        let password = args["password"] as? String ?? ""
        Task {
            do {
                mqttConnectionOptions = try await AuthManager.getMqttConfigForXS(
                    url: url, clientId: clientId, username: username, password: password
                )
                resolve(true)
            } catch {
                reject("getMqttConfigForXS(): \(error)", nil, nil)
            }
        }
    }

    @objc func connect(
        _ resolve: @escaping RCTPromiseResolveBlock,
        rejecter reject: @escaping RCTPromiseRejectBlock
    ) {
        if self.mqttConnectionOptions == nil {
            return reject("connect(): No MqttConfig set. Call register() first.", nil, nil)
        }
        Task {
            do {
                try await codingStation.connect(mqttConnectionOptions!)
                resolve(true)
            } catch {
                reject("connect(): \(error)", nil, nil)
            }
        }
    }

    @objc func write(
        _ resolve: @escaping RCTPromiseResolveBlock,
        rejecter reject: @escaping RCTPromiseRejectBlock
    ) {
        Task {
            do {
                try await codingStation.write()
                resolve(true)
            } catch {
                reject("write(): \(error)", nil, nil)
            }
        }
    }

    @objc func disconnect(
        _ resolve: @escaping RCTPromiseResolveBlock,
        rejecter reject: @escaping RCTPromiseRejectBlock
    ) {
        codingStation.disconnect()
        resolve(true)
    }
}

File: android/src/main/java/com/evva/xesar/abrevva/reactnative/AbrevvaCodingStation.kt
package com.evva.xesar.abrevva.reactnative

import android.app.Activity
import android.content.Intent
import android.view.View
import com.evva.xesar.abrevva.auth.AuthManager
import com.evva.xesar.abrevva.cs.CodingStation
import com.evva.xesar.abrevva.mqtt.MqttConnectionOptionsTLS
import com.facebook.react.ReactPackage
import com.facebook.react.bridge.ActivityEventListener
import com.facebook.react.bridge.NativeModule
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.uimanager.ReactShadowNode
import com.facebook.react.uimanager.ViewManager
import kotlinx.coroutines.DelicateCoroutinesApi
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.async
import kotlinx.coroutines.runBlocking
import java.net.URL
import java.util.Collections
import kotlin.jvm.Throws


class AbrevvaCodingStation(reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext),
    ActivityEventListener
{
    init {
        reactContext.addActivityEventListener(this)
    }

    private val codingStation = CodingStation(reactContext)
    private var mqttConnectionOptionsTLS: MqttConnectionOptionsTLS? = null

    @ReactMethod
    fun register(options: ReadableMap, promise: Promise) {
        val clientId = options.getString("clientId") ?: ""
        val username = options.getString("username") ?: ""
        val password = options.getString("password") ?: ""
        val url: URL
        try {
           url = URL(options.getString("url") ?: "")
        } catch (e:Exception){
            return promise.reject(Throwable("register(): invalid url"))
        }
        mqttConnectionOptionsTLS = AuthManager.getMqttConfigForXS(url, clientId, username, password)
        promise.resolve(true)
    }

    @ReactMethod
    fun connect(promise: Promise) {
        if (mqttConnectionOptionsTLS == null) {
            return promise.reject(Exception("connect(): No MqttCredentials present. call register() first"))
        }
        runBlocking {
            try {
                codingStation.connect(mqttConnectionOptionsTLS!!)
            } catch (e: Exception) {
                promise.reject(e)
            }
            promise.resolve(null)
        }
    }

    @ReactMethod
    fun write(promise: Promise) {
        runBlocking {
            codingStation.startTagReader(currentActivity!!,10_000) {
                promise.resolve(true)
            }
        }
    }

    @ReactMethod
    fun disconnect(promise: Promise) {
        codingStation.disconnect()
    }

    /// ActivityEventListener
    override fun onActivityResult(p0: Activity?, p1: Int, p2: Int, p3: Intent?) {
    }

    override fun onNewIntent(p0: Intent?) {
        codingStation.onHandleIntent(p0)
    }


    /// ReactContextBaseJavaModule
    override fun getName(): String {
        return NAME
    }

    companion object {
        const val NAME = "AbrevvaCodingStation"
    }
}
